### PR TITLE
o/snapstate: prevent install hang if prereq install fails

### PR DESCRIPTION
**Pull request by MiguelPires**
This PR prevents an install from hanging when its prerequisite has a duplicate install in the same change that fails. The following situation would trigger the bug: 
* `snap install A B` (where A depends on B)
* B's independent (i.e., not as a prerequisite) install fails
* A's installation starts and installs B as a prerequisite
* B's install checks for conflicts and fails, because the conflict checking code looks for a task affecting B that belongs to a change with tasks to be done (the previous install of B meets the criteria)
* The conflict error causes the prerequisite task to be retried after a timeout which causes the install be stuck in this loop, since the conflict is never "fixed"

One caveat: the check just skips without returning an error so A's install will proceed even though the other installed also failed (and B isn't installed). However, returning an error didn't seem right because there might be a situation where the install of the prerequisite conflicts with another install which hasn't completed but will succeed. Since the content provider is a soft dependency I thought it was more correct to just skip

### CODE REVIEW

File: overlord/snapstate/snapstate_install_test.go
Comment (mardy) [inline overlord/snapstate/snapstate_install_test.go:4052]: Does this mean that this instruction will always take 3 seconds to complete? If so, shouldn't we make the timeout much smaller?

(this is what I assume from the "won't converge" comment above, at least)

**Review by mardy — APPROVED** [review]
LGTM, I have just minor comments in the code. But on the other hand I don't grasp the high-level picture of this, so please wait for someone else's review before merging :-)

File: overlord/snapstate/snapstate_install_test.go
Comment (MiguelPires) [reply overlord/snapstate/snapstate_install_test.go:4047]: Yes, I would expect there's more than one task in the installation. The check just asserts there's something to be done for the install and not that there's a specific number of tasks since that'd be more brittle and IMO unnecessary for this case

File: overlord/snapstate/snapstate_install_test.go
Comment (MiguelPires) [reply overlord/snapstate/snapstate_install_test.go:4052]: Yes, I wasn't sure of how small to make it while ensuring it's enough for it to pass even in constrained environments. wdyt, 1 second?

File: overlord/snapstate/handlers.go
Comment (pedronis) [inline overlord/snapstate/handlers.go:321]: nitpick, this extra whitespace is unusual

**Review by pedronis — APPROVED** [review]
thanks, one nitpick

File: overlord/snapstate/snapstate_install_test.go
Comment (mardy) [reply overlord/snapstate/snapstate_install_test.go:4047]: OK

File: overlord/snapstate/snapstate_install_test.go
Comment (mardy) [reply overlord/snapstate/snapstate_install_test.go:4052]: I'd rather say 10 ms, but of course we should verify that's enough :-)

Unit tests should run quickly; if they don't, it means we need to add more mocking.

It seems weird, though, that we have an operation (Settle) which takes a timeout, but does not provide a channel for us to know when it's done...

Comment (pedronis) [reply overlord/snapstate/snapstate_install_test.go:4052]: > It seems weird, though, that we have an operation (Settle) which takes a timeout, but does not provide a channel for us to know when it's done...

I don't understand this comment, Settle is synchronous, if it returns without errors it's done

Comment (mardy) [reply overlord/snapstate/snapstate_install_test.go:4052]: > > It seems weird, though, that we have an operation (Settle) which takes a timeout, but does not provide a channel for us to know when it's done...
> 
> I don't understand this comment, Settle is synchronous, if it returns without errors it's done

Maybe there's a misunderstanding from my side then: I asked if that instruction, in the context of the test, will always take three seconds to complete, and the answer was yes. Can you confirm that?

If it is so, why cannot the `Settle()` operation return as soon as it's either failed or completed?

Comment (pedronis) [reply overlord/snapstate/snapstate_install_test.go:4052]: that's not true, the timeout it's a timeout, it will fail if it takes more than 3s, it doesn't mean it will always run for 3 sec

Comment (mardy) [reply overlord/snapstate/snapstate_install_test.go:4052]: Ah, OK, but then I'm fine with the current version of the code. :-)

File: overlord/snapstate/snapstate_install_test.go
Comment (stolowski) [inline overlord/snapstate/snapstate_install_test.go:4024]: Nitpick, I'd use the word "Conflict" in the test name, "Conf" may get confused with config-related stuff.

**Review by stolowski — APPROVED** [review]
LGTM, thanks, one nitpick.

File: overlord/snapstate/snapstate_install_test.go
Comment (MiguelPires) [reply overlord/snapstate/snapstate_install_test.go:4052]: @mardy sorry, the comment I left above the settle call is wrong and mislead you. Just to clarify, I thought the conflicting task would prevent the task runner from converging because the task never stopped being in `DoStatus`. However, that's not true because the task has no handler so it's ignored by the task runner

**Comment by mvo5** [comment]
Should we tag this for 2.52.1 or is it something that happens so rarely that it's fine to just have in 2.53?

**Comment by MiguelPires** [comment]
@mvo it should be quite rare, I think it's fine to have it in 2.53

**Pull request merged**
